redfish-core/lib/oem/ibm/usb_code_update.hpp

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include "async_resp.hpp"
+#include "dbus_utility.hpp"
+#include "redfish_util.hpp"
+
+#include <variant>
+
+namespace redfish
+{
+
+static constexpr const char* usbCodeUpdateObjectPath =
+    "/xyz/openbmc_project/control/service/phosphor_2dusb_2dcode_2dupdate";
+
+/**
+ * @brief Get the service name of the path
+ *
+ * @param[in] aResp     Shared pointer for generating response message.
+ * @param[in] path      The D-Bus Object path
+ * @param[in] handler   Call back
+ *
+ * @return None.
+ */
+template <typename Handler>
+inline void getServiceName(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
+                           const std::string& path, Handler&& handler)
+{
+    // Map of service name to list of interfaces
+    using MapperServiceMap =
+        std::vector<std::pair<std::string, std::vector<std::string>>>;
+
+    // Map of object paths to MapperServiceMaps
+    using MapperGetSubTreeResponse =
+        std::vector<std::pair<std::string, MapperServiceMap>>;
+
+    crow::connections::systemBus->async_method_call(
+        [aResp, path,
+         handler{std::move(handler)}](const boost::system::error_code ec,
+                                      const MapperGetSubTreeResponse& subtree) {
+            std::string service{};
+            if (ec)
+            {
+                BMCWEB_LOG_DEBUG << "DBUS response error, ec: " << ec.value();
+                messages::internalError(aResp->res);
+                return;
+            }
+
+            for (const auto& [objectPath, serviceMap] : subtree)

```

Comment:
Shouldn't you have some more checks here like: https://github.com/openbmc/bmcweb/blob/0fda0f12bb9ae0604a083dfae419be38a1418913/redfish-core/lib/systems.hpp#L1380

Reply:
> 

The objectPath will be checked at line 50.
An objectPath name may belong to multiple services, so there is no need to check the size of serviceMap.

Reply:
Done. Added some checks.